File: src/kt/PhantasyIslandKotlinRemoteControl/ControlCommand.kt
package PhantasyIslandKotlinRemoteControl

import java.util.concurrent.Executors
import java.util.concurrent.Future

open class AirplaneController(
    keyName: String,
    typeName: String,
    updateTimestamp: Long,
    status: AirplaneFlyStatus,
    cameraFront: String?,
    cameraDown: String?
) : AirplaneCore(keyName, typeName, updateTimestamp, status, cameraFront, cameraDown) {

    private var count: Int = 1
    private var sendCmdFn: (String) -> Any = { s -> HttpLayer.sendCmd(s) }
    private val executor = Executors.newFixedThreadPool(10)

    fun useFastMode(fastMode: Boolean = true, futureMode: Boolean = true) {
        if (!futureMode) {
            if (fastMode) {
                sendCmdFn = { s -> HttpLayer.sendCmdVolatile(s) }
            } else {
                sendCmdFn = { s -> HttpLayer.sendCmd(s) }
            }
        } else {
            if (fastMode) {
                sendCmdFn = { s -> executor.submit<Any> { HttpLayer.sendCmdVolatile(s) } }
            } else {
                sendCmdFn = { s -> executor.submit<Any> { HttpLayer.sendCmd(s) } }
            }
        }
    }

    private fun nextCount(): Int {
        count += 2
        return count
    }

    private fun prepareCommand(command: String): String {
        return "$keyName ${nextCount()} $command"
    }

    private fun sendCommand(command: String): Any {
        return sendCmdFn(prepareCommand(command))
    }

    fun mode(mode: Int) {
        airplaneMode(mode)
    }

    fun takeoff(high: Int): Any {
        return sendCommand("takeoff $high")
    }

    fun land(): Any {
        return sendCommand("land")
    }

    fun emergency(): Any {
        return sendCommand("emergency")
    }

    fun up(distance: Int): Any {
        return sendCommand("up $distance")
    }

    fun down(distance: Int): Any {
        return sendCommand("down $distance")
    }

    fun forward(distance: Int): Any {
        return sendCommand("forward $distance")
    }

    fun back(distance: Int): Any {
        return sendCommand("back $distance")
    }

    fun left(distance: Int): Any {
        return sendCommand("left $distance")
    }

    fun right(distance: Int): Any {
        return sendCommand("right $distance")
    }

    fun goto(x: Int, y: Int, h: Int): Any {
        return sendCommand("goto $x $y $h")
    }

    fun flip(direction: String): Any {
        return sendCommand("flip $direction 1")
    }

    fun flipForward() {
        flip("f")
    }

    fun flipBack() {
        flip("b")
    }

    fun flipLeft() {
        flip("l")
    }

    fun flipRight() {
        flip("r")
    }

    fun rotate(degree: Int): Any {
        return sendCommand("rotate $degree")
    }

    fun cw(degree: Int): Any {
        return sendCommand("cw $degree")
    }

    fun ccw(degree: Int): Any {
        return sendCommand("ccw $degree")
    }

    fun high(high: Int): Any {
        return sendCommand("high $high")
    }

    fun speed(speed: Int): Any {
        return sendCommand("setSpeed $speed")
    }

    fun led(r: Int, g: Int, b: Int): Any {
        return sendCommand("light $r $g $b")
    }

    fun bln(r: Int, g: Int, b: Int): Any {
        return sendCommand("bln $r $g $b")
    }

    fun rainbow(r: Int, g: Int, b: Int): Any {
        return sendCommand("rainbow $r $g $b")
    }

    fun airplaneMode(mode: Int): Any {
        return sendCommand("airplane_mode $mode")
    }

    fun stop(): Any {
        return hover()
    }

    fun hover(): Any {
        return sendCommand("hover")
    }
}

File: src/ex/lib/phantasy_island_remote_control/control_command.ex
defmodule PhantasyIslandRemoteControl.AirplaneController do
  @moduledoc """
  无人机控制，包含控制单个无人机的所有指令
  """

  alias PhantasyIslandRemoteControl.HttpLayer

  # 在 Elixir 中，由于是函数式语言，我们将 controller 作为一个状态传给函数
  # 或者将 key_name 作为第一个参数

  defp next_count(count), do: count + 2

  defp prepare_command(key_name, count, command) do
    "#{key_name} #{next_count(count)} #{command}"
  end

  defp send_cmd(key_name, count, command, opts) do
    full_cmd = prepare_command(key_name, count, command)
    case Keyword.get(opts, :fast_mode, false) do
      true -> HttpLayer.send_cmd_volatile(full_cmd)
      false -> HttpLayer.send_cmd(full_cmd)
    end
  end

  def takeoff(key_name, count, high, opts \\ []), do: send_cmd(key_name, count, "takeoff #{high}", opts)
  def land(key_name, count, opts \\ []), do: send_cmd(key_name, count, "land", opts)
  def emergency(key_name, count, opts \\ []), do: send_cmd(key_name, count, "emergency", opts)
  def up(key_name, count, distance, opts \\ []), do: send_cmd(key_name, count, "up #{distance}", opts)
  def down(key_name, count, distance, opts \\ []), do: send_cmd(key_name, count, "down #{distance}", opts)
  def forward(key_name, count, distance, opts \\ []), do: send_cmd(key_name, count, "forward #{distance}", opts)
  def back(key_name, count, distance, opts \\ []), do: send_cmd(key_name, count, "back #{distance}", opts)
  def left(key_name, count, distance, opts \\ []), do: send_cmd(key_name, count, "left #{distance}", opts)
  def right(key_name, count, distance, opts \\ []), do: send_cmd(key_name, count, "right #{distance}", opts)

  def goto(key_name, count, x, y, h, opts \\ []), do: send_cmd(key_name, count, "goto #{x} #{y} #{h}", opts)

  def flip(key_name, count, direction, opts \\ []), do: send_cmd(key_name, count, "flip #{direction} 1", opts)

  def rotate(key_name, count, degree, opts \\ []), do: send_cmd(key_name, count, "rotate #{degree}", opts)
  def cw(key_name, count, degree, opts \\ []), do: send_cmd(key_name, count, "cw #{degree}", opts)
  def ccw(key_name, count, degree, opts \\ []), do: send_cmd(key_name, count, "ccw #{degree}", opts)

  def high(key_name, count, high, opts \\ []), do: send_cmd(key_name, count, "high #{high}", opts)
  def speed(key_name, count, speed, opts \\ []), do: send_cmd(key_name, count, "setSpeed #{speed}", opts)

  def led(key_name, count, r, g, b, opts \\ []), do: send_cmd(key_name, count, "light #{r} #{g} #{b}", opts)
  def bln(key_name, count, r, g, b, opts \\ []), do: send_cmd(key_name, count, "bln #{r} #{g} #{b}", opts)
  def rainbow(key_name, count, r, g, b, opts \\ []), do: send_cmd(key_name, count, "rainbow #{r} #{g} #{b}", opts)

  def airplane_mode(key_name, count, mode, opts \\ []), do: send_cmd(key_name, count, "airplane_mode #{mode}", opts)
  def hover(key_name, count, opts \\ []), do: send_cmd(key_name, count, "hover", opts)
end
